### PR TITLE
feat: optimize health status updates

### PR DESCRIPTION
**Pull request by A1Liu**

### CODE REVIEW

File: server/utils/channel.ts [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+			this.reader = [resolveWrapper, reject];
+		});
+	}
+
+	destroy(error?: any): void {
+		if (this.isDestroyed) {
+			// destroying twice is probably an error; however, we'd then risk some
+			// scenario where we try to destroy to clean-up resources and then crash.
+			// better to just make sure a call to destroy can only happen once
```
Comment (karimsa) [inline server/utils/channel.ts:94]: That's fine, but probably want to at least log that it happened

File: server/utils/channel.ts [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+		}
+
+		const nextData = this.buffer.shift();
+		if (nextData) {
+			return nextData;
+		}
+
+		return new Promise<T | DESTROYED | TIMEOUT>((resolve, reject) => {
+			if (timeout === undefined) {
+				this.reader = [resolve, reject];
```
Comment (karimsa) [inline server/utils/channel.ts:72]: Lol why not an object? the `this.reader?.[1]?.()` is hard to read

**Review by karimsa — CHANGES_REQUESTED** [review]

File: server/utils/channel.ts (hunk 2 above)
Comment (A1Liu) [reply server/utils/channel.ts]: it is now an object

File: server/utils/channel.ts (hunk 1 above)
Comment (A1Liu) [reply server/utils/channel.ts:97]: done

File: server/services/service-builds.ts [hunk 3]
```diff
@@ -1,39 +1,130 @@
...
+
+				newServices.push(service);
+			}
+		}
+
+		// console.log(
+		// 	'xxx watchTask',
+		// 	newServices.map((s) => s.name),
+		// 	oldServices.map((s) => s.config.name),
+		// );
```
[the remark this answers is not in the document]
Comment (A1Liu) [reply server/services/service-builds.ts]: done

**Pull request merged**